std/btree.asm

```diff
@@ -0,0 +1,153 @@
+            CmpResult::Less => Option::Some(NodeSearchResult::InChild(i)),
+            CmpResult::Equal => Option::Some(NodeSearchResult::InNode(i)),
+            CmpResult::Greater => Option::None,
+        });
+        unwrap_or_else(r, || NodeSearchResult::InChild(std::array::len(items)))
+    };
+
+    enum InsertResult<K, V> {
+        Split((K, V), BTree<K, V>, BTree<K, V>),
+        Updated(BTree<K, V>)
+    }
+    let<K, V> insert: BTree<K, V>, (K, V), (K, K -> CmpResult) -> InsertResult<K, V> =
+        |b_tree, (k, v), cmp| match b_tree {
+            BTree::Leaf(items) => insert_into_leaf(items, (k, v), cmp),
+            BTree::Inner(items, children) =>
+                match search_in_node(items, k, cmp) {
+                    NodeSearchResult::InNode(i) => InsertResult::Updated(BTree::Inner(array_set(items, i, (k, v)), children)),
+                    NodeSearchResult::InChild(i) => match insert(children[i], (k, v), cmp) {
+                        InsertResult::Updated(child) => InsertResult::Updated(BTree::Inner(items, array_set(children, i, child))),
+                        InsertResult::Split((k1, v1), left, right) =>
+                            insert_into_inner(items, children, (k1, v1), i, left, right),
+                    }                    
+                }
+        };
+    let<K, V> insert_into_leaf: (K, V)[], (K, V), (K, K -> CmpResult) -> InsertResult<K, V> = |items, (k, v), cmp| {
+        let (new_items, _) = items_insert(items, (k, v), cmp);
+        if std::array::len(new_items) <= super::max_items {
+            InsertResult::Updated(BTree::Leaf(new_items))
+        } else {
+            split_leaf(new_items)
+        }
+    };
+    let<K, V> split_leaf: (K, V)[] -> InsertResult<K, V> = |items| {
+        let split = (std::array::len(items) - 1) / 2;
+        let left = sub_array(items, 0, split);
+        let right = sub_array(items, split + 1, std::array::len(items) - split - 1);
+        InsertResult::Split(items[split], BTree::Leaf(left), BTree::Leaf(right))
+    };
+    let<K, V> insert_into_inner:
+        (K, V)[], BTree<K, V>[], (K, V), int, BTree<K, V>, BTree<K, V> -> InsertResult<K, V> =
+        |items, children, (k, v), i, left, right| {
+            let new_items = sub_array(items, 0, i) + [(k, v)] + sub_array(items, i, std::array::len(items) - i);
+            // We replace the old children[i] by [left, right]
+            let new_children =
+                sub_array(children, 0, i) +
+                [left, right] +
+                sub_array(children, i + 1, std::array::len(children) - i - 1);
+            if std::array::len(new_items) <= super::max_items {
+                InsertResult::Updated(BTree::Inner(new_items, new_children))
+            } else {
+                split_inner(new_items, new_children)
+            }
+        };
+    let<K, V> split_inner: (K, V)[], BTree<K, V>[] -> InsertResult<K, V> = |items, children| {
+        let split = (std::array::len(items) - 1) / 2;
```

Comment:
Same suggestions as for `split_leaf`, assert that the lengths are as expected and then use `super::max_items` instead of `std::array::len`.